### ISSUE TITLE
bar chart: monthly scale off by 1 into the future, and 11/23 is shown as 0/24

### ISSUE THREAD

**Issue by gitisgreat2023** [ioBroker/ioBroker.echarts#492]
Monthly timestamps at 23:59 are displayed one month into the future. E.g. 2023 Sept 30th 23:59 is shown as Oct 2023.
Nov 30 23:59 is actually even shown as 0/2024 (not 12/2023, or at least off by one as 12/2023).

Somehow the whole time axis of the bar chart is buggy. I'm willing to fix the code, can give someone a head-start where the issue might be hidden, in which lines?

**Comment by SilasMueller** [comment]
Isnt this exactly the same you already posted? https://github.com/ioBroker/ioBroker.echarts/issues/475

**Comment by Norb1204** [comment]
Some hints:

- I assume that the month starts with 0 instead of 1. This is probably the shift of one month.
- It looks like that the calculation is based on 30 days per month. But months have days between 28 and 31.
- This concerns not only months, also days that are shifted by 1 into the future (Probably also hours).

I believe that the general problem is that everything starts from 0 instead of 1 and the calculation of the months is always with 30 days.

**Comment by gitisgreat2023** [comment]
> Isnt this exactly the same you already posted? #475

no, I see a shift by 2 months here, not by one day.
November is for example displayed as 0/2024 (january)

**Comment by jh2k19** [comment]
The same proplem here.
![image](https://github.com/ioBroker/ioBroker.echarts/assets/51288287/659d74ce-e492-42f3-9ea6-b150a3552a47)
![image](https://github.com/ioBroker/ioBroker.echarts/assets/51288287/87ac2032-675e-48f7-a28b-b7998f1ddcd8)

**Comment by jh2k19** [comment]
Sorry, I missed something there. Under Appearance there is a parameter X-Label-offset. This is the desired solution for me.
![image](https://github.com/ioBroker/ioBroker.echarts/assets/51288287/94f21527-03c7-4f73-9cb4-b7ed001346d3)

**Comment by gitisgreat2023** [comment]
Yeah that works. Though the mouse over still shows the wrong value. I have a new issue, two months are counted as one (February 2024).

**Comment by Norb1204** [comment]
A comment exists in the code "ChartModel.js", maybe someone can take a look at it and assist the developer a bit.

<img width="706" alt="ChartModel" src="https://github.com/ioBroker/ioBroker.echarts/assets/132662785/3de78e92-bb2a-4211-b4a1-f4f94fab35bc">

**Comment by jh2k19** [comment]
Hier ein Codeschnipsel den ich mal mit ChatGPT ertellet habe.
Ich weis aber nicht wie man den Einbait bzw. möchte hier am Code auf Git nichts falsch oder kaputt machen.


function daysInMonth(year, month) {
    return new Date(year, month + 1, 0).getDate();
}

function calculateMonths(start, end) {
    start = new Date(start);
    end = new Date(end);
    
    // Set to the beginning of the next month
    end.setDate(1);
    end.setHours(0);
    end.setMinutes(0);
    end.setSeconds(0);
    end.setMilliseconds(0);
    end.setMonth(end.getMonth() + 1);
    
    let totalDays = 0;
    let current = new Date(start.getTime());
    
    while (current < end) {
        totalDays += daysInMonth(current.getFullYear(), current.getMonth());
        current.setMonth(current.getMonth() + 1);
    }
    
    let months = totalDays / 30;
    return Math.round(months);
}

let start = new Date('2023-01-15').getTime(); // Beispiel-Startdatum
let end = new Date('2023-06-17').getTime(); // Beispiel-Enddatum


Erklärung des Codes:

daysInMonth(year, month) Funktion:

Diese Funktion gibt die tatsächliche Anzahl der Tage im angegebenen Monat zurück.
Sie verwendet den Trick, dass der Tag 0 des nächsten Monats der letzte Tag des aktuellen Monats ist.
calculateMonths(start, end) Funktion:

Diese Funktion berechnet die Anzahl der Monate zwischen start und end.
Der end Zeitpunkt wird auf den Beginn des nächsten Monats gesetzt.
Eine Schleife summiert die Anzahl der Tage in jedem Monat zwischen start und end.
Schließlich wird die Gesamtanzahl der Tage durch 30 geteilt, um die Anzahl der Monate zu berechnen, und das Ergebnis wird gerundet.
Dieser Ansatz berücksichtigt die tatsächliche Anzahl der Tage in jedem Monat und liefert eine genauere Berechnung.

let option = {};
option.count = calculateMonths(start, end);

console.log(option.count);

**Comment by Erwin40** [comment]
> Sorry, I missed something there. Under Appearance there is a parameter X-Label-offset. This is the desired solution for me. ![image](https://private-user-images.githubusercontent.com/51288287/334149084-94f21527-03c7-4f73-9cb4-b7ed001346d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5MDkwMDQsIm5iZiI6MTcyMTkwODcwNCwicGF0aCI6Ii81MTI4ODI4Ny8zMzQxNDkwODQtOTRmMjE1MjctMDNjNy00ZjczLTljYjQtYjdlZDAwMTM0NmQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI1VDExNTgyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1NmZjZTQwNzJmYWQwZDFiZDI2NTQwYmMwN2M2YjkyZThkNmY2NDZmZjJjMTViNDA3M2QwNzZkMjU3YjE4YTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tt7TbLnIc-AEkMjOl6ZxdLJaCiH1r8VUjX3JyL9A6R4)

Das ist ziemlicher Unsinn.
Dann stimmt das Datum der Cursor Position nicht mehr mit dem im ToolTip angezeigten Datum überein :-(
